README
# MEAN Stack REST API template using TypeScript
This is a simple way to get started in a mean project.

This project includes:
- express
- mongoose
- dotenv
- cors
- bcrypt
- helmet
- debug
- cookie-parser
- http-errors
- jsonwebtoken
- mongo-store
- morgan
- auth routes (signup & login)
- auth middleware (authorization skeleton, authenticationRequired and apiToken required)
- basic user model
- env.example with every used variable


CHANGELOG.md
# CHANGELOG

## v0.1.1
- Added nodemailer
- Light Refator
    - routes, renamed to controllers
    - Every file indicates which type of data is inside (e.g. auth.model.ts && auth.controller.ts && auth.mailer.ts vs auth.ts && auth.ts && auth.ts)
- Mailers implementation (ruby-on-rails-like)
- Helpers implementation (ruby-on-rails-like)
- Auth improvements:
    - Forgotten password recovery
    - Welcome email
    - Email address confirmation
    - Email address change
- User model improvements
    - name, referred (usefull if you want to implement a referral-system feature), and emailConfirmed fields
- Session model for managing request to change data in users
- Added user controller
    - User profile retrieving endpoint
    - User profile updating endpoint
    - User password change endpoint

## v0.1.0
- Initial project structure
A simple express app with a authentication sub app and several packages one usually add to an api project as a boilerplate.
- Helmet to aim security issues
- Very basic user model
- Signup and Login endpoints
- Auth middleware to restrict api-token and jwt based requests
- Auth middleware barebones for require a certain level for the user
- Fallback errors (404 & 500)


src/routes/auth.ts
import express, { Request, Response, Router } from 'express';
import { User } from '../models/user';
import { hash, compare } from 'bcrypt';
import { sign } from 'jsonwebtoken';

export const auth = express();

export const authRouter = Router();

authRouter.post('/login', (req: Request, res: Response) => {
    const { email, password } = req.body;
    if (email && password) {
        User.findOne({ email }).then(user => {
            if (user) {
                compare(password, user.password).then(encrypted => {
                    const { JWT_SECRET } = process.env;
                    const token = sign({
                        user_id: user._id,
                        email: user.email,
                    }, JWT_SECRET as any as string, { expiresIn: '30d' });
                    const { password, ...data} = (user as any)._doc;
                    const payload = {
                        token,
                        code: 201,
                        message: 'User authenticated',
                        data,
                    };
                    res.status(201).json(payload);
                },
                (error) => {
                    res.status(403).json({ code: 403, message: error.message });
                });
            } else {
                res.status(404).json({ code: 404, message: 'User not found' });
            }
        })
    } else {
        res.status(500).json({ code: 500, message: 'Invalid data provided' });
    }
});

authRouter.post('/signup', (req: Request, res: Response) => {
    const { email, password } = req.body;
    if (email && password) {
        hash(password, 10).then(encrypted => {
            if (encrypted) {
                const data: {email: string, password:string} = {
                    email,
                    password: encrypted
                };

                const user = new User(data);
                user.save().then(
                    () => {
                        res.status(201).json({ code: 201, message: 'User created.' });
                    },
                    (error) => {
                        res.status(500).json({ code: 500, message: error.message });
                    }
                )
            } else {
                res.status(500).json({ code: 500, message: 'Couldn\'t encrypt data' });
            }
        }).catch(error => {
            res.status(500).json({ code: 500, message: error.message });
        });
    } else {
        res.status(500).json({ code: 500, message: 'Invalid data provided' });
    }
});

auth.use('/', authRouter);


src/mailers/auth.mailer.ts
import { config } from 'dotenv';
import { sendHtmlEmail } from './base.mailer';
config();

const {
	MAIL_BUTTON_HOST
} = process.env;

export const sendPasswordRecoveryEmail = (payload: { to: string, _id: string, update_token: string }) => {
	const { to, _id, update_token } = payload;
	return sendHtmlEmail({
		from: '"Otoniel (kenliten) Reyes Galay" <otoniel@otonielreyes.com>',
		to,
		subject: "Password recovery",
		html: `To change your password click the link bellow.<br/>
        <a href="${MAIL_BUTTON_HOST}/auth/restore-password/${_id}/${update_token}">Reset my password now</a>
        `
	});
}

export const sendEmailConfirmEmail = (payload: { to: string, _id: string, confirmation_token: string }) => {
	const { to, _id, confirmation_token } = payload;
	return sendHtmlEmail({
		from: '"Otoniel (kenliten) Reyes Galay" <otoniel@otonielreyes.com>',
		to,
		subject: "Email address confirmation",
		html: `To confirm your email address click the link bellow.<br/>
        <a href="${MAIL_BUTTON_HOST}/user/confirm-email/${_id}/${confirmation_token}">Confirm my email now</a>
        `
	});
}

export const sendEmailChangeConfirmEmail = (payload: { to: string, _id: string, confirmation_token: string }) => {
	const { to, _id, confirmation_token } = payload;
	return sendHtmlEmail({
		from: '"Otoniel Reyes Galay" <otoniel@otonielreyes.com>',
		to,
		subject: "Email address change confirmation",
		html: `To change and confirm your new email address click the link bellow.<br/>
        <a href="${MAIL_BUTTON_HOST}/user/change-email/${_id}/${confirmation_token}">Confirm my email now</a>
        `
	});
}

export const sendWelcomeEmail = (to: string) => {
	return sendHtmlEmail({
		from: '"Otoniel (kenliten) Reyes Galay" <otoniel@otonielreyes.com>',
		to,
		subject: "Welcome to AwesomeApp",
		html: `<h1>Welcome to AwesomeApp</h1>
		<p>It's a pleasure to have you here with us.</p>
		<p>If you need something, just email me. I'll be pleased to help.</p>
		`
	});
}

src/middleware/auth.middleware.ts
import { NextFunction, Request, Response } from "express";
import { User } from "../models/user";
import { verify } from "jsonwebtoken";

export function requiresAuthentication(req: Request, res: Response, next: NextFunction) {
    const token = req.body.token || req.query.token || req.headers['x-access-token'] || req.headers['authorization']?.slice(6);

    if (!token) {
        return res.status(403).json({ code: 403, message: 'Missing access token.' });
    }

    try {
        const { JWT_SECRET } = process.env;
        if (JWT_SECRET) {
            const decoded = verify(token, JWT_SECRET);
            (req as any).user = decoded;
        } else {
            return res.status(500).json({ code: 500, message: 'Misconfiguration' });
        }
    } catch (err: any) {
        return res.status(403).json({ code: 403, message: err.message });
    }

    return next();
}

export function requiresAuthorization(req: Request, res: Response, next: NextFunction) {
}

export function requiresAPIToken(req: Request, res: Response, next: NextFunction) {
    const token = req.header('x-api-token');
    if (token) {
        User.findOne({ api_token: token }).then(
            user => {
                if (user) {
                    (req as any).user = user;
                    next();
                } else {
                    res.status(403).json({ message: 'Invalid api token.', code: 403 });
                }
            },
            error => {
                res.status(403).json({ message: error.message, code: 403 });
            }
        );
    } else {
        res.status(403).json({ message: 'Missing api token.', code: 403 });
    }
}

src/controllers/auth.controller.ts
import express, { Request, Response, Router } from 'express';
import { hash, compare } from 'bcrypt';
import { config } from 'dotenv';
import { sendPasswordRecoveryEmail, sendWelcomeEmail } from '../mailers/auth.mailer';
import { Session } from '../models/session.model';
import { User } from '../models/user.model';
import { generateToken, createAuthToken } from '../helpers/auth.helper';
config();

export const authController = express();

export const authRouter = Router();

authRouter.post('/signup', (req: Request, res: Response) => {
    const { email, password } = req.body;
    if (email && password) {
        hash(password, 10).then(encrypted => {
            if (encrypted) {
                const data: {email: string, password: string, api_token: string} = {
                    email,
                    password: encrypted,
                    api_token: generateToken(64),
                };

                const user = new User(data);
                user.save().then(
                    async (data) => {
                        const session = new Session({ user_id: user._id });
                        session.save();
                        const token = createAuthToken(user);
                        const { password, ...userData} = (user as any)._doc;
                        const payload = {
                            token,
                            code: 201,
                            message: 'User created',
                            data: userData,
                        };
                        await sendWelcomeEmail(email);
                        res.status(201).json(payload);
                    },
                    (error) => {
                        res.status(500).json({ code: 500, message: error.message });
                    }
                )
            } else {
                res.status(500).json({ code: 500, message: 'Couldn\'t encrypt data' });
            }
        }).catch(error => {
            res.status(500).json({ code: 500, message: error.message });
        });
    } else {
        res.status(500).json({ code: 500, message: 'Invalid data provided' });
    }
});

authRouter.post('/login', (req: Request, res: Response) => {
    const { email, password } = req.body;
    if (email && password) {
        User.findOne({ email }).then(user => {
            if (user) {
                compare(password, user.password).then(result => {
                    if (result) {
                        const token = createAuthToken(user);
                        const { password, ...data} = (user as any)._doc;
                        const payload = {
                            token,
                            code: 201,
                            message: 'User authenticated',
                            data,
                        };
                        res.status(201).json(payload);
                    } else {
                        res.status(400).json({
                            code: 400,
                            message: 'Wrong password.'
                        });
                    }
                },
                (error) => {
                    res.status(403).json({ code: 403, message: error.message });
                });
            } else {
                res.status(404).json({ code: 404, message: 'User not found' });
            }
        });
    } else {
        res.status(500).json({ code: 500, message: 'Invalid data provided' });
    }
});

authRouter.post('/forgot-password', (req: Request, res: Response) => {
    const update_token = generateToken();
    if (req.body.email) {
        User.findOne({ email: req.body.email }).then(user => {
            if (user) {
                const { _id } = (user as any)._doc;
                Session.updateOne({ user_id: _id }, {
                    passwordUpdateRequestToken: update_token,
                    date: new Date(),
                    expires: new Date(Date.now() + (1000*60*60))
                }).then((count) => {
                    if (count) {
                        sendPasswordRecoveryEmail({
                            to: req.body.email,
                            _id,
                            update_token
                        }).then(() => {
                            res.status(201).json({
                                code: 201,
                                message: 'Password recovery request created.'
                            });
                        }).catch(error => {
                            res.status(500).json({
                                code: 500,
                                message: error.message,
                            });
                        });
                    } else {
                        res.status(500).json({
                            code: 500,
                            message: 'Can\'t save the request'
                        });
                    }
                });
            }
        }).catch(error => {
            res.status(404).json({
                code: 404,
                message: 'User not found'
            });
        })
    } else {
        res.status(400).json({
            code: 400,
            message: 'No email provided'
        });
    }
});

authRouter.post('/password-reset', (req: Request, res: Response) => {
    if (req.body.password && req.body._id && req.body.token) {
        Session.findOne({
            user_id: req.body._id,
            passwordUpdateRequestToken: req.body.token
        }).then(session => {
            if (session) {
                if (new Date < new Date(`${session.expires}`)) {
                    Session.updateOne({ user_id: req.body._id }, {
                        expires: new Date(),
                        passwordUpdateRequestToken: '',
                    });
                    hash(req.body.password, 10).then(password => {
                        User.updateOne({ _id: req.body._id }, {
                            password
                        }).then(async (count) => {
                            if (count) {
                                const user = await User.findOne({ _id: req.body._id });
                                if (!user) {
                                    return res.status(500).json({
                                        code: 500,
                                        message: 'Password updated'
                                    });
                                }
                                const token = createAuthToken(user);
                                const { password, ...data} = user;
                                const payload = {
                                    token,
                                    code: 201,
                                    message: 'Password updated',
                                    data,
                                };
                                res.status(201).json(payload);
                            } else {
                                res.status(500).json({
                                    code: 500,
                                    message: 'Password unchanged'
                                })
                            }
                        }).catch(error => {
                            res.status(500).json({
                                code: 500,
                                message: error.message,
                            });
                        });
                    }).catch(error => {
                        res.status(500).json({
                            code: 500,
                            message: error.message,
                        });
                    });
                } else {
                    res.status(400).json({
                        code: 400,
                        message: 'Expired token'
                    });
                }
            } else {
                res.status(400).json({
                    code: 400,
                    message: 'Bad request'
                });
            }
        })
    } else {
        res.status(400).json({
            code: 400,
            message: 'Bad request'
        });
    }
});

authController.use('/', authRouter);